Topic 2: Chi-Square Tests

### Task 2

Part of assignments for the Machine Learning and Statistics modulel of the Higher Diploma in Science in Data Analytics course at ATU 2023

Winter 23/24

Author: Jarlath Scarry

### Task 2

### Chi-square test

> Consider the below contingency table based on a survey asking respondents whether they prefer coffee or tea and whether they prefer plain or chocolate biscuits. Use scipy.stats to perform a chi-squared test to see whether there is any evidence of an association between drink preference and biscuit preference in this instance.

![images/task2_image1.PNG](images/task2_image1.PNG)



### Assumptions for Chi-square test

When you choose to analyse your data using a chi-square test for independence, you need to make sure that the data you want to analyse "passes" two assumptions. You need to do this because it is only appropriate to use a chi-square test for independence if your data passes these two assumptions. If it does not, you cannot use a chi-square test for independence. These two assumptions are:

Assumption #1:
Your two variables should be measured at an ordinal or nominal level (i.e., categorical data). You can learn more about ordinal and nominal variables in our article: Types of Variable.

Assumption #2:
Your two variable should consist of two or more categorical, independent groups. Example independent variables that meet this criterion include gender (2 groups: Males and Females), ethnicity (e.g., 3 groups: Caucasian, African American and Hispanic), physical activity level (e.g., 4 groups: sedentary, low, moderate and high), profession (e.g., 5 groups: surgeon, doctor, nurse, dentist, therapist), and so forth. These independent groups should not have any overlap. For example Tea or coffee drinker, Well sometimes I drink Tea, and sometimes Coffee. This is not accepted. The choice should be either or.

In the section, Procedure, we illustrate the SPSS Statistics procedure to perform a chi-square test for independence. First, we introduce the example that is used in this guide.

[[1]](#ref1) (statistics.laerd.com Chi-Square Test for Association using SPSS Statistics Oct 2023)


We have Two variables which are cross tabulated on the table. One variable is Drinks with a choice of Tea or Coffee on the Verticle axis on the Left, and the other Variable is Bisciuts with ca choice of Chocolate or Plain on the horizontal axis at the top.

### Chi-Square Tests

    
### Laerd Statistics Chi-Square Test for Independence


### Data for the test
Data is generated from the results table above.  It is done in a few steps below using pandas, and generated results are shuffled to make them look more as expected.

In [45]:
import pandas as pd
import random

import scipy.stats as ss
from scipy.stats.contingency import crosstab


coffee_chocolate = [['Coffee','Chocolate']]*43
coffee_plain = [['Coffee','Plain']]*57
tea_chocolate = [['Tea','Chocolate']]*56
tea_plain = [['Tea','Chocolate']]*56

raw_data = coffee_chocolate + coffee_plain + tea_chocolate + tea_plain
#shuffle the data
random.shuffle(raw_data)
# Zip the list - make the rows columns and the columns rows
# Interchange the outer and inner lists
drink, biscuit = list(zip(*raw_data))  #2 lists, one with deing and one with biscuits.

# create a data frame
df = pd.DataFrame({'drink': drink, 'biscuit': biscuit})

df  # df generated from the table in the question.
df.to_csv('data\survey_task2.csv') 

## Contingency Table

Contingency tables are used in statistics to summarize the relationship between several categorical variables. In our example, The Contingency table between the two variables "Drinks " and "Biscuits" is a Frequency table of these variables presented simultaneously. A chi-squared test conducted on a contingency table can test whether or not a relationship exists between variables. These effects are defined as relationships between rows and columns.

[[2]](#ref2)(stackoverflow.com, How to understand the chi square contingency table, Oct 2023)

The Contingency Table function produces a table of the joint distribution of two categorical variables. This technique is often used to analyze survey data such as in our small survey.

[[3]](#ref3)(scipy.org, scipy.stats.contingency.crosstab, Oct 2023)

scipy.stats.contingency.crosstab

> This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table observed. [[4]](#ref4) (Wikipedia.org, Contingency table, Oct 2023) 

> The expected frequencies are computed based on the marginal sums under the assumption of independence; see scipy.stats.contingency.expected_freq. The number of degrees of freedom is (expressed using numpy functions and attributes).

It returns a table of counts for each possible unique combination in the argument


In [46]:
#cross = pd.crosstab(index=df['drink'], columns=df['biscuit'])
#cross

In [47]:
# Perform Crosstabs Contingency.

cross = ss.contingency.crosstab(df['drink'], df['biscuit'])
# Show.
cross



CrosstabResult(elements=(array(['Coffee', 'Tea'], dtype=object), array(['Chocolate', 'Plain'], dtype=object)), count=array([[ 43,  57],
       [112,   0]]))

In [48]:
first, second = cross.elements
first, second

(array(['Coffee', 'Tea'], dtype=object),
 array(['Chocolate', 'Plain'], dtype=object))

In [49]:
cross.count

array([[ 43,  57],
       [112,   0]])

In [50]:
import scipy
print('SciPy version:', scipy.__version__)

SciPy version: 1.11.3


In [51]:
df[df['drink'] == first[0]]

drink    biscuit
0    Coffee      Plain
2    Coffee      Plain
4    Coffee      Plain
5    Coffee  Chocolate
10   Coffee  Chocolate
..      ...        ...
202  Coffee      Plain
204  Coffee      Plain
205  Coffee  Chocolate
207  Coffee      Plain
209  Coffee  Chocolate

[100 rows x 2 columns]

####  Statistical Test

> Chi-square test of independence of variables in a contingency table. [[5]](#ref5)(scipy.org, scipy.stats.chi2_contingency, Oct 2023) 

> This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table [1] observed. The expected frequencies are computed based on the marginal sums under the assumption of independence; see scipy.stats.contingency.expected_freq. The number of degrees of freedom is (expressed using numpy functions and attributes):


#### In our case the paramaters are:

    The array
    
    Correction value.  When set to false this prevents the function from applying the Yates correction
  
> In statistics, Yates's correction for continuity (or Yates's chi-squared test) is used in certain situations when testing for independence in a contingency table. It aims at correcting the error introduced by assuming that the discrete probabilities of frequencies in the table can be approximated by a continuous distribution (chi-squared). In some cases, Yates's correction may adjust too far, and so its current use is limited. [[6]](#ref6)(wikipedia.org, Yates's correction for continuity, Oct 2023) https://en.wikipedia.org/wiki/Yates%27s_correction_for_continuity 

> Like scipy.stats.chisquare, this function computes a chi-square statistic; the convenience this function provides is to figure out the expected frequencies and degrees of freedom from the given contingency table.

    dof degrees of freedom.   
   

In [52]:
#the contingency table
cross.count

# Do the statistics. Just do them.
Chi2ContingencyResult = ss.chi2_contingency(cross.count, correction=False) ##chi2_contingency is the function that does the chi2 test for independence. 

# Show.
Chi2ContingencyResult

Chi2ContingencyResult(statistic=87.31664516129032, pvalue=9.24664373812942e-21, dof=1, expected_freq=array([[73.11320755, 26.88679245],
       [81.88679245, 30.11320755]]))

### Results

#### Expected results?
What results would we expect to see? Well we can see the "expected frequency" data already on the results table. This can also be quickly calculate this using the `expected_freq` function.  This returns the expected frequencies based on the marginal sums of the table. It shows a results table the same shape as the original table observed. 

The expected results show what result we should expect to see in a table of independent groups. In this case the expected results can be seen in the table below. For example we would expect to see around 73 people to respond saying they would like Coffee with a Chocolate biscuit.
       
[[5]](#ref5)(scipy.org, scipy.stats.chi2_contingency, Oct 2023)  

In [53]:
import numpy
expected = Chi2ContingencyResult.expected_freq
numpy.savetxt("data\expected_task2.csv", expected, delimiter=",")
title = 'Expected result'
expected_df = pd.DataFrame(expected,columns=['Chocolate','Plain'],index=['Coffee', 'Tea'])
expected_df = expected_df.style.set_caption(title)
expected_df

#### Actual survey results
<img src="images/task2_image1.PNG" align="left"/>

### Actual results

The function returns a results  `object` containing the following attributes:

  statistic float:  The test statistic. = `statistic=87.31664516129032`

  pvalue float: The p-value of the test. = `pvalue=9.24664373812942e-21`

  dof int: The degrees of freedom.
       This is generally calculated as `dof` = (number of rows - 1) * (number of columns - 1) =`dof=1`

  expected_freqndarray, same shape as observed
        The expected frequencies, based on the marginal sums of the table.

[[5]](#ref5)(scipy.org, scipy.stats.chi2_contingency, Oct 2023) 

### interpreting the results

In the Chi-squared test for independence we are testing if there is a difference in the proportions across different cattegories.

Is the choice of a plain or chocolate biscuit dependant on the drink chosen? Does knowing what Drink a person chooses tell us anything about what biscuit the might like? Looking at the data it appears that somone chhosing Tea is more likley to choose a Chocolate biscuit, but we can check this with the Chi-squared test for independence. 

If we assume at the outset that there is no difference between whether a person likes Tea or Coffee with either Plain or Chocolate biscuits, and take this to be k0, or the NULL hypothesis. If this is true what are the chances that the sample data we have would fit that hypothesis? We need to set the treshold for this hypothesis. If we pass the treshold we reject the NULL hypothesis. 

So we can take it that a p-value of less than 5% (p-value<0.05) means the result is statistically significant. In otherwords we reject the NULL hypothesis and accept that the sample data is representitive of the overall population. That if a person chooses Tea, they are likley to have a chocolate biscuit. Our p-value is far lower than 5% so we should accept the alternative hypothesis. In other words our assunption that the groups are independent is false based on out survey.

### Conclusion

 For a Chi-square test, a p-value that is less than or equal to your significance level indicates there is sufficient evidence to conclude that the observed distribution is not the same as the expected distribution. You can conclude that a relationship exists between the categorical variables. [[7]](#ref7)(statisticsbyjim.com, Chi-Square Test of Independence and an Example, Oct 2023) 
 
Our p-value is far lower than 5% so we should accept the alternative hypothesis. In other words our assunption that the groups are independent is false based on out survey.

So we should conclude that if somone chooses Tea to drink they are more likley to have a Chocolate biscuit.
 
The p-value is the evidence against a null hypothesis. The smaller the p-value, the strong the evidence that you should reject the null hypothesis. [[8]](#ref8)(stackoverflow.com, how to understand the chi square contingency table, Oct 2023) 



### Acknowledgments

Ian McLoughlin ATU.ie MLAS lecture notes. For much of the code and inspiration for the rest.

### References

[1]<a id='ref1'></a> (statistics.laerd.com Chi-Square Test for Association using SPSS Statistics Oct 2023) https://statistics.laerd.com/spss-tutorials/chi-square-test-for-association-using-spss-statistics.php

[2]<a id='ref2'></a> (stackoverflow.com How to understand the chi square contingency table, Oct 2023)
https://stackoverflow.com/questions/52692315/how-to-understand-the-chi-square-contingency-table

[3]<a id='ref3'></a>(scipy.org, scipy.stats.contingency.crosstab, Oct 2023)
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.contingency.crosstab.html

[4]<a id='ref4'></a>(Wikipedia.org, Contingency table, Oct 2023) https://en.wikipedia.org/wiki/Contingency_table 

[5]<a id='ref5'></a>(scipy.org, scipy.stats.chi2_contingency, Oct 2023) https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html

[6]<a id='ref6'></a>(wikipedia.org, Yates's correction for continuity, Oct 2023) https://en.wikipedia.org/wiki/Yates%27s_correction_for_continuity

[7]<a id='ref7'></a>(statisticsbyjim.com, Chi-Square Test of Independence and an Example, Oct 2023) https://statisticsbyjim.com/hypothesis-testing/chi-square-test-independence-example/#:~:text=For%20a%20Chi%2Dsquare%20test,exists%20between%20the%20categorical%20variables. 

[8]<a id='ref8'></a>(stackoverflow.com, how to understand the chi square contingency table, Oct 2023) https://stackoverflow.com/questions/52692315/how-to-understand-the-chi-square-contingency-table


END